syntax = "proto3";

package teton.proto;

message Mesh {
    int32 id = 1;
    string name = 2;
    repeated Layer layers = 3;
}

message Layer {
    int32 id = 1;
    float height = 2; // Z position
    float line_width = 3;
    float line_thickness = 4;

    repeated LayerPart parts = 5;
}

message LayerPart {
    int32 id = 1;

    repeated PrintArea areas = 3;
}

message PrintArea {
    enum Type {
        UnknownType = 0;
        Exterior = 1;
        Wall = 2;
        Skin = 3;
        Infill = 4;
        Hole = 5;
    }
    
    enum InfillPattern {
        UnknownPattern = 0;
        Grid = 1;
        Triangle = 2;
        Cubic = 3;
    }

    int32 id = 1;
    int32 inside = 2; // id of PrintArea this PrintArea is inside of
    //bytes points = 3; // polygon
    repeated int32 points = 3; // polygon

    Type type = 4;

    float orientation = 5; // skin or infill orientation
    float infill_line_spacing = 6;
    InfillPattern infill_pattern = 7;
}

